w3af/plugins/audit/open_api_auth.py

```diff
@@ -0,0 +1,372 @@
+Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+
+"""
+import copy
+
+import w3af.core.data.kb.knowledge_base as kb
+import w3af.core.controllers.output_manager as om
+
+from w3af.core.controllers.plugins.audit_plugin import AuditPlugin
+from w3af.core.data.constants import severity
+from w3af.core.data.kb.vuln import Vuln
+
+
+class open_api_auth(AuditPlugin):
+    """
+    Checks if REST API endpoints authenticate incoming requests
+    as defined in their API specification.
+    See details in the plugin description below.
+
+    Here is a list of possible enhancements:
+      * Consider sending requests with other HTTP methods,
+        and check that 401 or 405 error code is returned
+        On the one hand, it may not look too good if we just replace the method,
+        and keep the rest of the request untouched (headers, payload, etc).
+        But on the other hand, the application is supposed to check HTTP method
+        in the very beginning, and reject requests if the method is not supported.
+        If the API spec says a particular method is supported,
+        then the check should expect 401
+        If the API spec says a particular method is not supported,
+        then the check should expect 405 (preferred) or 401
+        (should it only expect 405 in this case?).
+      * Allow a user to specify response codes that the plugin should expect.
+        Currently the plugin expects only 401.
+        We can introduce 'expected_code_regex' configuration parameter
+        which set a regex for expected response codes.
+      * If the plugin is updated to send other HTTP methods (see above),
+        we may want to introduce another configuration parameter
+        to set expected response codes in this case.
+      * Severity of identified issues may depend on response codes
+        returned by the server.
+
+    :author: Artem Smotrakov (artem.smotrakov@gmail.com)
+    :author: Andres Riancho (andres.riancho@gmail.com)
+    """
+
+    def __init__(self):
+        AuditPlugin.__init__(self)
+        self._spec = None
+        self._expected_codes = [401]
+
+    def audit(self, freq, orig_response, debugging_id):
+        """
+        Check if an API endpoint requires authentication according to its API spec.
+
+        :param freq: A FuzzableRequest to the API endpoint.
+        :param orig_response: The HTTP response associated with the fuzzable request.
+        :param debugging_id: A unique identifier for this call to audit().
+        """
+        # The check works only with API endpoints.
+        # crawl.open_api plugin has to be called before.
+        if not self._is_api_spec_available():
+            om.out.information("Could not find an API specification, skip")
+            return
+
+        # The API spec must define authentication mechanisms.
+        if not self._has_security_definitions_in_spec():
+            om.out.information("The API spec has no security definitions, skip")
```

Comment:
Change message to `The Open API spec has no security definitions, not running open_api_auth checks.`

Also, this should be printed only once per each open-api specification document w3af finds.